### DISCUSSION TITLE
AddLumpedPort definitions and general definitions

### DISCUSSION THREAD

**Discussion by WidderGit** [thliebig/openEMS-Project#231] [General]
Hello,
I have a few specific questions I could use some help with and a general question
Volker recently helped me with an example.  When I got to the port section of the program I am struggling with the "excite" definition.  
port1 = FDTD.AddLumpedPort(1, 50, [start], [stop], 'z', excite=1.0, priority=150)
When excite = 1.0, what is "1.0" is this a voltage or a current or something else?  The literature just says this is the "port excitation amplitude"
The literature also seems to say that whatever this value is it needs to be defined along a direction.  In this case it looks like the z direction.  Assuming this is a voltage, why is a direction needed?  
I have also think I read where an excitation this can be in V/length, which I don't follow, perhaps the length is the size of the port?

General question (and I don't want to sound unappreciative).  When I search the openEMS sight I am struggling with some of the definitions (like above) as the descriptions seem to be missing some information.  Is there an alternate source that provides more description about the inputs to the software?  
I also have no idea what **kw means... but I am sure it will be obvious when I find out.

Thanks everyone
Marc

**Comment by VolkerMuehlhaus** [comment]
> Assuming this is a voltage, why is a direction needed?

What you have used here is a port, which has two terminals. The port excitation (let's assume a voltage) is applied between the two terminals. The port is in z-direction and there is a metal conductor at the top side and another metal conductor at the bottom. 

Inside the port, there is a source with the internal resistor, as well as voltage and current probe. You can inspect these items if you switch the respective elements on and off in AppCSXCAD, which shows all of them individually.

Hope this helps?

<img width="711" alt="port_openems" src="https://github.com/user-attachments/assets/d2bc4672-3873-4c57-9295-2a642c71aa12">

**Reply by WidderGit** [reply]
Volker,
Yes it does help, 
When you set excite =1, I assume that this is 1V across the 50 ohm port and basically the resistance of the rest of the loop is based on whatever the material properties are set to?  So if this is correct, and the resistance of the rest of the conductors (total) is 10 ohms, I would expect a "steady state" current of ~.0167A?

Thanks
Marc

**Reply by VolkerMuehlhaus** [reply]
Hi Marc,

I'm not sure if it is 1V and not even sure if it's a voltage source. For the FDTD solver that I use at work, it is a 1A current source in parallel to the port resistor, which means 50V excitation from the 50 Ohm source. 

For the openEMS example that I gave to, I see rather small port voltages of 1.3mV or so as excitation (named "u1 incident" in the plot), not sure how that is calculated from the "1.0" excitation. There is a linear dependency, but I don't know the internals, and usually I don't care because this is all linear and can be scaled to other values as you wish.

Sorry I have no better answer, most of my FDTD work is with a commercial solver and openEMS is very similar in many aspects, but not identical.

Best regards
Volker

**Reply by WidderGit** [reply]
Volker
Looks like the steady state current is ~2.6e-5 A, so the excitation is likely not in Amps.  If the excitation is in Volts, then the I calculated the resistance of the system to be ~3.85e4 Ohms.  The literature says CSX.AddMetal creates a perfect electrical conductor, so I assume no resistance... so the calculated value doesn't fit.  All a bit confusing, perhaps someone else will chime in.
Thanks, 
Marc

**Reply by VolkerMuehlhaus** [reply]
Hi Marc, 

yes, I hope someone can chime in. 

But be careful with the time signals at this small time scale - this is not the static DC value that you see for 50 Ohm source and load. We see the instantaneous impedance as the signal (step in voltage) propagates along the line. It takes some round trips between source and load until we see the steady state value expected for 50 Ohm source and load. Before that we see values that depend on local line impedance etc., which was the original purpose of this example (TDR analysis to determine local deviation in line impedance).

As I wrote above, according to the voltage plots the incident signal is around 1.3mV (not 1V). The port voltage then converges to that value after some round trips. Current converges to around 0.025 mA and thus impedance is approximately 1.3/0.025 = 52 Ohm. That is expected, because we have the load at port 2 which determines the DC (steady state) current. That port impedance is 50 Ohm, so there is some small error in the numbers read from the plot, but the overall picture is correct.

@thliebig Thorsten, can you explain what source signal (voltage or current unit) we expect for excitation value of "1"?

**Reply by WidderGit** [reply]
Thanks Volker,
Marc

**Reply by thliebig** [reply]
I would need to have a closer look but If I remember correctly this is directly the electric field inside the excitation box.
Thus the total voltage and current of the port depends on the port geometry too.
But as you said already, usually you do not really care about the value itself, only about relations or use it to normalize.
The only issue I see with the current approach is if you have multiple different type and/or shaped ports excited at the same time and except/want them to have the same port voltage...

**Reply by WidderGit** [reply]
Thorsten,
A few questions/comments.
For the immediate problem of interest, if I want 1A of steady state current, should I just adjust the excitation value until I get the desired current?  
In the future, I'd like to be able to drive the circuit with specific time varying voltage levels.  Can you recommend a source where I can determine how to calculate the voltage I am applying, based on the port geometry and the excitation (electric field) value?
I am confused about the last comment (and this question may have been directed to Volker?).   Are you recommending that the second port be replaced with a simple conductor?  This is related to the second question as I don't understand how the ports interact/drive the circuit.
Thanks,
Marc

**Reply by thliebig** [reply]
For almost all your question I'm just asking myself.. WHY? Why do you want to excite a specific current or voltage? 
Since everything is linear you can run the simulation with whatever voltage or current happens to happen and then scale every result value to whatever voltage or current (port-) reference you could ever want without running another simulation...

**Comment by WidderGit** [comment]
Thorsten,
I think my answer to your question/comment has 3 parts.
1) I'd like to eventually model the behavior of circuits with known transient source voltages.  If I know how to input the voltage at the port directly, without trial and error, I can more efficiently solve the problems.
2) To ensure I am getting correct results, and to make sure I am not missing something, I think it is important that I have a rudimentary idea of how the software works (algorithms?).  In this particular case I want to understand what "excitation" means and how the software got from an excitation of 1 to the observed current/voltage.  So, in summary, I want to learn.
3) I think you essentially answered one of my 3 questions.  Which is if I want to check the simulated results against the standard equation, B = (mu0)I/2(pi)r, I can just use the resulting steady state current to make the test calculation.  Since everything is linear, if I really want 1 A of current, I can just scale the excitation until I get something close to 1A.  Please correct me if this is wrong.

For context, I have experience and training with explicit and implicit CFD and structural analysis/software.  I am very comfortable with these codes due to this experience.  Since I have not been trained in FDTD, and have no experience with openEMS, I am very interested in what the program is doing, in order to feel more comfortable with the results.

Hope this helps you understand my questions.
Thanks
Marc

**Reply by thliebig** [reply]
1) what kind of transient voltages? Because if the signals are much longer than a few nano seconds simulation times explode very quickly. Usually it is prefered to create e.g. a touchstone file and do signal analysis in external circuit simulations
2)  There is very much literature about the FDTD method and even books and simple examples to implement your own very simple 1D-FDTD to get a better idea about the details.
3) sounds about right... But I would not call it steady state... usually you excite a broadband Gaussian pulse and then do a DFT for the desired frequency on the resulting data.

**Comment by WidderGit** [comment]
Can anyone provide insight on how ports work, and how to generate either a specific voltage vs time profile or current vs time profile with the "excite" instruction?
Thanks
Marc

**Reply by thliebig** [reply]
Simply put a port is just a collection of simpler building blocks. In openEMS this usually (not always) includes and excitation (meaning electric or magnetic fields are imprinted either soft or hard in the simulation domain), an optional (usually 50Ohms) resistor and some voltage and current probes.  The excite argument of a port usually is just an on/off and scaling value. As said above it usually effects the amplitude of the excitation (see above).

**Comment by WidderGit** [comment]
Thorsten,
Sorry for the late reply, I have been studying the FDTD method.
I am looking for times around 50 ns.  

As far as I can tell, the only line in the code that applies a non-zero excitation is:
port1 = FDTD.AddLumpedPort(1, 50, [start], [stop], 'z', excite=1.0, priority=150)

I don't see where this line defines the excitation as an electric or magnetic field or whether it is hard or soft, can you tell me how this is accomplished?  From what I have discovered it looks like the units for an electric filed source are in V/m.  Since the port is 5mm long, this would make the voltage input 5mV, for an excitation of 1.

It also looks like there is another form of excitation called a "total field/scatter field" source.  Can openEMS apply this type of source?

Thanks
Marc

**Reply by WidderGit** [reply]
Volker,
I may have found some answers.  When I look at the voltage vs time in port1, it starts at .005V, which is equivalent to 1V/m times the 5mm length of the port.  It appears this voltage is just an initial condition, not constant (as I initially thought).   The current appears to keep going in the circuit (at least up to the end of the solve).  When I multiply the end current times the 50 ohm port value, I get very close to the ~.0013 "steady state" voltage across each port.  This is nearly identical to the reported port voltages.  It appears that "excite" defaults to V/m...?
Anyway, assuming I am correct, perhaps this will help someone understand the software a little better.  Also, perhaps you were trying to tell me this earlier and I didn't get it at the time.
Marc

**Reply by VolkerMuehlhaus** [reply]
> It appears that "excite" defaults to V/m...?

Sounds good to me, this seems to be in line with the comments by Thorsten above, who is the author of openEMS. 
_"this is directly the electric field inside the excitation box"_


> The current appears to keep going in the circuit (at least up to the end of the solve). When I multiply the end current times the 50 ohm port value, I get very close to the ~.0013 "steady state" voltage across each port. This is nearly identical to the reported port voltages. 

This sounds like a misunderstanding - the port voltage is the **superposition** of the **port excitation** (e.g. gaussian pulse or step in voltage) and **residual pulses** that originate from the excitation and keep travelling between possible discontinuities at the respective propagation speed.

By looking at a very short time scale, and because your testcase line is not matched to the port impedance, we see a lot of these reflections and the instantaneous voltage at the port varies a lot in the initial time frame when the pulse travels along the line and gets reflected. 

I don't know if you understand that stuff, how the instantaneous voltage changes when the signal travelling along the line hits a different line impedance (different line width)? This topic is related to TDR (time domain reflectometry) and I had showcased one example here:
https://github.com/thliebig/openEMS-Project/discussions/177#discussioncomment-8132900

Perhaps you are overwhelmed with RF and signal integrity effects due to the very short pulses in your model, which are short/fast compared to the propagation time? Not sure if you need that at all, and if perhaps a simulation in frequency domain would be a much nicer (more efficient) solution. That's what we EM simulation professionals usually do: simulate the model with sufficient bandwith in frequency domain, and extract the frequency domain S-parameters. These can then be used as a "complete" black box description of the device under test, and you can now excitate that black box in **circuit simulation** (not EM!) with all sorts of fancy transient signals in time domain, to see the black box response. But notice that we don't do direct EM simulation with these fancy time domain signals, we just get the S-parameters from gaussian pulse FDTD to get wideband S-parameters, that's all we do in EM. The rest is circuit simulation using that black box S-parameter data. I use Keysight ADS for this taks, but there are other commercial tools which can do the same thing.

So your "brute force" direct EM solution is somewhat unusual, although it might be possible if you know the tools well.

**Reply by WidderGit** [reply]
Volker,
Yes, going in I didn't realize that openEMS is used mainly for RF type applications, whereas what I am doing is a bit different.
However, to optimize what I am doing I'll eventually need to pay attention to S parameters etc...

Since I figured out that the voltage at the port is just an initial condition, I am attempting to figure out how to apply a transient voltage.  I found the code below (as a starting point) but, as I suspected, this does not work with Python.  Do you happen to know what the equivalent code would be for Python?

% Define the port excitation
port1 = FDTD.AddLumpedPort(1, 50, [start], [stop], 'z', 'excite', @(t) sin(2*pi*1e9*t));

Thanks
Marc

**Reply by VolkerMuehlhaus** [reply]
Marc, you seem to misunderstand the workflow. The port voltage (or to be more precise, the excitation inside the port) is not "just an initial condition". This misunderstanding might occur due to the request for a time domain plot, and the related **choice** (!) of a step in voltage as the excitation.

Your idea to use a single frequency sine wave would be rather inefficient. In FDTD we can use a **gaussian pulse** which corresponds to a wide frequency spectrum, all in one pulse and one single simulation run. Using that gaussian pulse we excite a wide bandwidth simulatenously (!) and the port voltages and currents are then postprocessed (FFT'ed) in the CalcPort() function to get the frequency domain response. You will find this in the very basic openEMS examples like MSL_NotchFilter.py

Excitation defined as **gaussian pulse** to cover a certain bandwidth 
```
### Setup FDTD parameters & excitation function
FDTD = openEMS()
FDTD.SetGaussExcite( f0, fc)
```

where f0 = center frequency and fc =20dB cutoff frequency --> bandwidth is 2*fc

For postprocessing note that we now look at the "uf" components to get the voltage in **frequency** domain.

```
### Post-processing and plotting
f = linspace( 1e6, f_max, 1601 )
port1.CalcPort( Sim_Path, f, ref_impedance = 50)
port2.CalcPort( Sim_Path, f, ref_impedance = 50)

s11 = port1.uf_ref / port1.uf_inc
s21 = port2.uf_ref / port1.uf_inc
```

**Reply by WidderGit** [reply]
Volker,
I left out some context in an attempt to not over complicate the question, but I have added this now.
I think I understand the intent of the gaussian pulse and obtaining the s parameters but that is not what I am trying to solve.
I am attempting to apply custom voltage/time pulse and monitor the transient effect on the H field at a distance away from the "wire".
As a start I found an example that used a time varying sine wave, that was the only reason for the code I provided above.  When I attempted to apply the same code to the model you gave me, (as an excitation), it did not work (as expected) because it is not Python code.  I was not able to find an example in the openems literature for how to make a custom time varying excitation, so I was asking if you (or anyone else) knows what the code should be for the case that I found.  
Ultimately, I'll need to write a custom formula where the excitation will vary with the solution time.
Hope this explains what I am attempting to do.
Thanks
Marc

**Reply by VolkerMuehlhaus** [reply]
Yes, Marc, I understand that part and would solve it indirectly.

But anyway, you have your workflow in mind and that is absolutely posssible using a commercial EM solver (Empire XPU). Not sure if that is an option, or if you have a requirement to solve this using openEMS? 

I built something in Empire to test what is my understanding of your planned workflow:

1. Microstrip transmission line with ports on both ends, and one field probe (point) that samples in time domain

<img width="1005" alt="emp1" src="https://github.com/user-attachments/assets/a5e86425-e8d6-45b8-8e36-42c401d1a248">

2. Custom excitation file input data from file (format: time and amplitude)

<img width="911" alt="emp2" src="https://github.com/user-attachments/assets/478a1173-effd-4bfd-96b6-e7a03a5d70bb">

3. Time plot of excitation (et, max value is 1A) and port voltages at line input and output

<img width="838" alt="emp3" src="https://github.com/user-attachments/assets/3cda7bf2-bda9-41f4-b7ac-cacaa0ed3533">

Excitation is current (!) into the 50 Ohm port, and combined with the 50 Ohm input impedance of the line, we have 25 Ohm load at the 1A source -> 1A*25 Ohm = 25V port voltage
You can also see the electrical delay of the line, port 2 voltage is delayed compared to port 1

4. Time domain reponse of the H field at the defined probe location, x,y,z components

<img width="838" alt="emp4" src="https://github.com/user-attachments/assets/7bdc5080-28a7-4e08-8a93-be2a70051bf3">

So your workflow is possible. My openEMS experience is rather limited, and some of this might not be possible, but I hope the other will jump in to comment. 
If you have a budget you might consider to switch to the commercial Empire XPU solver, which offers support and full documentation. 

BR,
Volker

**Reply by WidderGit** [reply]
Volker,
I was not aware of Empire XPU and will definitely get in touch with them to see if I can afford a seat of this software.  I was mainly using openEMS as I was concerned that I would not be able to afford a commercial software.
You have truly gone above and beyond to help me with the project, so once again thanks for your input.
Marc

**Reply by VolkerMuehlhaus** [reply]
Hi Marc, 
just in case you get started with an Empire XPU demo, I have attached my example file here.
[Userexcite_fieldsample_v9.zip](https://github.com/user-attachments/files/17210110/Userexcite_fieldsample_v9.zip)

Regarding the custom excitation or sinusoidal excitation in openEMS: 
The port parameter is really just a number to control the (peak) amplitude. The shape (time response) is defined elsewhere, that is the code section with FDTD.SetGaussExcite(), FDTD.SetStepExcite(), FDTD.SetSinusExcite() or FDTD.SetCustomExcite

You can find some more detail in the comments in openEMS sources for Python
https://github.com/thliebig/openEMS/blob/master/python/openEMS/openEMS.pyx

The FDTD.SetCustomExcite() had been missing in the Python interface of openEMS, and was added recently. I have no experience with using that, because we usually use SetGaussExcite() for S-Parameter calculation ("all frequencies contained in one pulse").

BR, Volker

**Reply by WidderGit** [reply]
Volker,
I received a prompt response from Empire and unfortunately the cost is much higher than I can afford, but thanks for the input.
I'll see if I can figure out the Python code with the recommended reference material above.  I am new to Python as well as openEMS so getting the code correct is challenging.  If I figure it out, I'll post the solution as it may help others.
Thanks
Marc
